tendermint/src/serializers.rs

```diff
@@ -1,71 +1,199 @@
+//! ```
+//! use serde::{Serialize, Deserialize};
+//! use tendermint::serializers;
+//!
+//! #[derive(Serialize, Deserialize)]
+//! struct ByteTypes {
+//!
+//!     #[serde(with="serializers::bytes::hexstring")]
+//!     hexbytes: Vec<u8>,
+//!
+//!     #[serde(with="serializers::bytes::base64string")]
+//!     base64bytes: Vec<u8>,
+//!
+//!     #[serde(with="serializers::bytes::string")]
+//!     bytes: Vec<u8>,
+//!
+//! }
+//! ```
+//!
+//! Available serializers:
+//! i64                  <-> string:               #[serde(with="serializers::primitives::string")]
+//! u64                  <-> string:               #[serde(with="serializers::primitives::string")]
+//! std::time::Dureation <-> nanoseconds as string #[serde(with="serializers::timeduration::string")]
+//! Vec<u8>              <-> HexString:            #[serde(with="serializers::bytes::hexstring")]
+//! Vec<u8>              <-> Base64String:         #[serde(with="serializers::bytes::base64string")]
+//! Vec<u8>              <-> String:               #[serde(with="serializers::bytes::string")]
+//!
+//! Any type that has the "FromStr" trait can be serialized into a string with serializers::primitives::string.
+//!
 
 use crate::account::{Id, LENGTH};
 use crate::{block, Hash, Signature};
 use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 use std::str::FromStr;
-use std::time::Duration;
-use subtle_encoding::{base64, hex};
 
-/// Parse `i64` from a JSON string
-pub(crate) fn parse_i64<'de, D>(deserializer: D) -> Result<i64, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    String::deserialize(deserializer)?
-        .parse::<i64>()
-        .map_err(|e| D::Error::custom(format!("{}", e)))
-}
+/// Serialize/deserialize primitive types (i64, u64, etc)
+pub mod primitives {
 
-/// Serialize `i64` as a JSON string
-#[allow(clippy::trivially_copy_pass_by_ref)]
-pub(crate) fn serialize_i64<S>(value: &i64, serializer: S) -> Result<S::Ok, S::Error>
-where
-    S: Serializer,
-{
-    format!("{}", value).serialize(serializer)
-}
+    /// Serialize into string, deserialize from string
+    pub mod string {
+        use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 
-/// Parse `u64` from a JSON string
-pub(crate) fn parse_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    String::deserialize(deserializer)?
-        .parse::<u64>()
-        .map_err(|e| D::Error::custom(format!("{}", e)))
-}
+        /// Deserialize string into T
+        pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
+        where
+            D: Deserializer<'de>,
+            T: std::str::FromStr,
+            <T as std::str::FromStr>::Err: std::fmt::Display,
+        {
+            String::deserialize(deserializer)?
+                .parse::<T>()
+                .map_err(|e| D::Error::custom(format!("{}", e)))
+        }
```

Comment:
Is it only me or was it surprising to you as well that serder doesn't provide this?

Reply:
I was lucky, because I got to read Tony's code first. So I was prepared for it and I just chalked it up as yet another Rust curiosity. :)